### DISCUSSION TITLE
Support offline development

### DISCUSSION THREAD

**Discussion by monk233** [pocketbase/pocketbase#152] [Q&A] [answered]
When I tried to create an admin account with a correct mailbox, I received this error "must be a valid email address."
![image](https://user-images.githubusercontent.com/52564907/179383114-4366b446-df92-465f-85b5-5fc6d608397f.png)
When I put "amazingmonk@gmail.com" Replace with "amazingmonk@example.com" Successfully created when
![image](https://user-images.githubusercontent.com/52564907/179383186-f2b50ef5-4cfc-46c6-b5d6-2d536ba4c63c.png)
version 0.2.8

**Comment by ganigeorgiev** [comment]
Hi, thanks for reporting the issue but I cannot reproduce it.

Maybe the error is from a previous incorrect email attempt (we are clearing the errors on input unfocus/blur).

If you are still having troubles using the email address, please let me know and I'll investigate it further.

**Comment by monk233** [comment]
Thank you for having time to verify my question.

I tried to delete PD_ Data, but it still doesn't solve the problem.

![image](https://user-images.githubusercontent.com/52564907/179404373-567d82b0-75ef-4907-9a87-bc9b68a88161.png)

So I chose to pass first test@example.com This mailbox enters pocketbase.

After entering pocketbase, I tried to use an email field to create a new record, but I encountered the same problem. Except that the mailbox with the suffix @example can be successfully created, other mailboxes failed.

![image](https://user-images.githubusercontent.com/52564907/179404389-cfc3ade6-3dac-4532-b456-bd9f65a03ad6.png)

So I think we should start with the source code and check whether there is a problem with the email verification of the back-end service.

In the process of checking the source code, I found that the pocketbase authentication mailbox uses an authentication library ozzo validation
![image](https://user-images.githubusercontent.com/52564907/179404574-e0a737a8-2b84-40d3-b8c9-7e4d763dc195.png)
![image](https://user-images.githubusercontent.com/52564907/179404660-1f3d2e28-0c1d-4bf7-b5ce-43b749c53820.png)

So I went to check how ozzo validation verifies email.
When I see the definition of validation rules, I think I know why the email I entered fails to validate (because there is no MX record)
![image](https://user-images.githubusercontent.com/52564907/179404739-d5a190d1-9397-4359-a43a-a8a84fdbe973.png)

Then I continued to read the code of the validation rule and found out why the mailbox of @example passed the validation
![image](https://user-images.githubusercontent.com/52564907/179404912-fe851d17-4002-41ec-a761-b8ad5d469cac.png)

The above is the whole process of finding problems. I didn't run the whole project to verify my ideas.

Thank you again for checking my question.

**Comment by ganigeorgiev** [comment]
Yes, that's correct. The email validation requires the domain to have a valid MX record.
**Which means that it is also required to have an active internet connection** (probably I should mention that somewhere in the docs).

Are you testing the application offline?

**Comment by monk233** [comment]
Yes, I'm testing the application offline, and I usually use some mock data.

Is it possible to validate MX records as an option?

**Comment by ganigeorgiev** [comment]
I could change it to validate only the email format, but I'm not sure how useful this would be because there are other parts of the application that requires you to be connected to the internet (eg. the OAuth2 authentication, emails sending, connecting to external s3 storage, etc.).

I'll move the issue to discussions so the other users can also share their opinion whether adding an offline mode would be beneficial.

**Comment by traeblain** [comment]
@ganigeorgiev I was looking at using this on an internal system that has no need for external access, and with IT lockdowns it makes it impossible for this to work (i.e. if a service doesn't need access beyond the intranet, it doesn't get it).  It's not just offline, it's also behind a locked down IT network that makes using it problematic.  MX validation should be optional.

**Reply by ganigeorgiev** [reply]
Hm, networks with restricted outbound traffic is something I haven't considered before.
I don't want to make the MX validation optional because it will clutter the interface and complicate unnecessary the internals.
I'm more inclined to replace the MX validator with just email format checking, but I'll think a little more on this after the upcoming v0.6.0 release.

**Comment by ganigeorgiev** [comment, marked as answer]
@traeblain, @monk233 I've replaced the MX domain check with plain email format validator in the latest [v0.6.0 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.6.0) which should allow to use PocketBase (to some extend) in network restricted environments.

**Comment by jamauro** [comment]
> whether adding an offline mode would be beneficial

Curious to hear more about what this might enable. Would this be akin to an "offline-first" / "local-first" solution? That is, a user could still interact with an app even if they go offline temporarily and changes would be synced when they came back online. Or do you mean something else by offline mode?

**Reply by ganigeorgiev** [reply]
@jamauro 

>  Would this be akin to an "offline-first" / "local-first" solution?

No. This discussion was for the email validators whether they should apply a MX domain check or not. As mentioned in https://github.com/pocketbase/pocketbase/discussions/152#discussioncomment-3534061, from v0.6 we now check only the email format so that it doesn't cause issues in environments with restricted external network access.

Offline data sync may be added in the SDKs, but for now this is left for the developers to handle on their own if they really need it.

**Reply by jamauro** [reply]
> Offline data sync may be added in the SDKs

This would be a great addition

**Reply by rodydavis** [reply]
Working on an offline first SDK here:
https://github.com/rodydavis/pocketbase_drift
